aptos-move/framework/aptos-framework/sources/state_storage.move

```diff
@@ -51,14 +59,41 @@ module aptos_framework::state_storage {
                 bytes: 0,
             }
         });
-        move_to(aptos_framework, GasParameter {
-            usage: Usage {
-                items: 0,
-                bytes: 0,
-            }
+        move_to(aptos_framework, StorageGasParameter {
+            per_byte: 0,
+            per_item: 0,
         });
     }
 
+    /// The storage gas base follows a 1/4 arc curve and scale it to [0, SCALE_FACTOR].
+    fun calculate_gas(target: u64, current: u64): u64 {
+        let r = (target as u128);
+        let x = if (current < target) {(current as u128)} else {r};
+        let y = (isqrt(r * r - x * x) as u64);
+        (target - y) * SCALE_FACTOR / target
+    }
+
+    // Find the greatest number x that x * 2 <= n.
+    fun isqrt(n: u128): u128{
+        let x = n;
+        let c = 0;
+        let d = 1u128 << 126;
+        while (d > n) {
+            d = d >> 2;
+        };
+        while (d != 0) {
+            if (x >= c + d) {
+                x = x - (c + d);
+                c = (c >> 1) + d;
+            }
+            else {
+                c = c >> 1;
+            };
+            d = d >> 2;
+        };
+        c
+    }
```

Comment:
I guess I may want to move this to https://github.com/aptos-labs/aptos-core/pull/3172